x/ibc/handler.go

```diff
@@ -37,22 +38,75 @@ func NewHandler(k Keeper) sdk.Handler {
+			if err != nil {
+				return nil, err
+			}
+			res, cap, err := channel.HandleMsgChannelOpenInit(ctx, k.ChannelKeeper, portCap, msg)
+			cbs.ClaimCapability(ctx, cap, ibctypes.ChannelCapabilityPath(msg.PortID, msg.ChannelID))
+			return res, err
 
 		case channel.MsgChannelOpenTry:
-			return channel.HandleMsgChannelOpenTry(ctx, k.ChannelKeeper, msg)
+			cbs := k.PortKeeper.LookupModule(ctx, msg.PortID)
+			portCap, _ := cbs.GetCapability(ctx, ibctypes.PortPath(msg.PortID))
+			err := cbs.OnChanOpenTry(ctx, msg.Channel.Ordering, msg.Channel.ConnectionHops, msg.PortID, msg.ChannelID, msg.Channel.Counterparty, msg.Channel.Version, msg.CounterpartyVersion)
+			if err != nil {
+				return nil, err
+			}
+			res, cap, err := channel.HandleMsgChannelOpenTry(ctx, k.ChannelKeeper, portCap, msg)
+			cbs.ClaimCapability(ctx, cap, ibctypes.ChannelCapabilityPath(msg.PortID, msg.ChannelID))
+			return res, err
 
 		case channel.MsgChannelOpenAck:
-			return channel.HandleMsgChannelOpenAck(ctx, k.ChannelKeeper, msg)
+			cbs := k.PortKeeper.LookupModule(ctx, msg.PortID)
+			chanCap, _ := cbs.GetCapability(ctx, ibctypes.ChannelCapabilityPath(msg.PortID, msg.ChannelID))
+			err := cbs.OnChanOpenAck(ctx, msg.PortID, msg.ChannelID, msg.CounterpartyVersion)
+			if err != nil {
+				return nil, err
+			}
+			return channel.HandleMsgChannelOpenAck(ctx, k.ChannelKeeper, chanCap, msg)
 
 		case channel.MsgChannelOpenConfirm:
-			return channel.HandleMsgChannelOpenConfirm(ctx, k.ChannelKeeper, msg)
+			cbs := k.PortKeeper.LookupModule(ctx, msg.PortID)
+			chanCap, _ := cbs.GetCapability(ctx, ibctypes.ChannelCapabilityPath(msg.PortID, msg.ChannelID))
+			err := cbs.OnChanOpenConfirm(ctx, msg.PortID, msg.ChannelID)
+			if err != nil {
+				return nil, err
+			}
+			return channel.HandleMsgChannelOpenConfirm(ctx, k.ChannelKeeper, chanCap, msg)
 
 		case channel.MsgChannelCloseInit:
-			return channel.HandleMsgChannelCloseInit(ctx, k.ChannelKeeper, msg)
+			cbs := k.PortKeeper.LookupModule(ctx, msg.PortID)
+			chanCap, _ := cbs.GetCapability(ctx, ibctypes.ChannelCapabilityPath(msg.PortID, msg.ChannelID))
+			err := cbs.OnChanCloseInit(ctx, msg.PortID, msg.ChannelID)
+			if err != nil {
+				return nil, err
+			}
+			return channel.HandleMsgChannelCloseInit(ctx, k.ChannelKeeper, chanCap, msg)
 
 		case channel.MsgChannelCloseConfirm:
-			return channel.HandleMsgChannelCloseConfirm(ctx, k.ChannelKeeper, msg)
+			cbs := k.PortKeeper.LookupModule(ctx, msg.PortID)
+			chanCap, _ := cbs.GetCapability(ctx, ibctypes.ChannelCapabilityPath(msg.PortID, msg.ChannelID))
+			err := cbs.OnChanCloseConfirm(ctx, msg.PortID, msg.ChannelID)
+			if err != nil {
+				return nil, err
+			}
+			return channel.HandleMsgChannelCloseConfirm(ctx, k.ChannelKeeper, chanCap, msg)
+
+		// IBC packet msgs get routed to the appropriate module callback
+		case channel.MsgPacket:
+			cbs := k.PortKeeper.LookupModule(ctx, msg.Packet.DestinationPort)
```

Comment:
Rather than assuming portID == moduleName and using baseapp Router, we lookup module here in IBC handler and call appropriate callback

Reply:
This actually needs to be based on the channel capability, not the port capability